NEP: Add NEP 50 draft about fixing scalar promotion rules

seberg opened the pull request:
Note, this really is a very early draft.  For myself, I have not
even 100% made up my mind that this is actually the best approach.

However, to solicitate feedback and input, I would like to put it
out there (and have a nice rendered version).

For those mainly interested in the possible different approaches
(i.e. making decisions), the "Alternatives" section is probably most
interesting.

Please make a note of blatend errors or unnecessary repitition, but
don't be too surprised if you see them, I will not be surprised if
this needs to be redone from scratch eventually.

---

Artifact link: https://25381-908607-gh.circle-artifacts.com/0/doc/neps/_build/html/nep-0050-scalar-promotion.html   (Updated 2022-03-16)

(early) branch for testing: https://github.com/seberg/numpy/tree/weak-scalars  (Note that this branch currently does not attempt to give warnings, or implement the `uint8_arr + 1000 -> error` behavior, which is more complicated.  It should implement the general "weak" logic though.

seberg committed:
NEP: Add early draft for fixing scalar promotion rules

Note, this really is a very early draft.  For myself, I have not
even 100% made up my mind that this is actually the best approach.

However, to solicitate feedback and input, I would like to put it
out there (and have a nice rendered version).

For those mainly interested in the possible different approaches
(i.e. making decisions), the "Alternatives" section is probably most
interesting.

Please make a note of blatend errors or unnecessary repitition, but
don't be too surprised if you see them, I will not be surprised if
this needs to be redone from scratch eventually.

jni commented:
For what it's worth, your summary from the mailing list post:

> Roughly speaking, I have made a proposal under the 3 points:
> * NumPy scalars and NumPy arrays always behave the same.
> * A NumPy array always respects the dtype
> * A Python scalar is "weak" so that uint8_arr + 3 returns a uint8_arr

reflects exactly how I voted, and I think is the best approach, *specifically* because (as far as I can tell) Python scalars are indistinguishable from *literals* and I would expect `uint8_array + 3` to remain uint8. So a big :+1: from me! Thanks @seberg for tackling the Hard Problems! :blush:

seberg commented:
Thanks, I do encourage you to have a brief look at the alternatives (if you have not done before), because some stuff may also just be a way to try for fewer backcompat breaks, even if we don't think it is the _purest_ approach...

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+Note that the examples apply just as well for operations like multiplication,
+addition, or comparisons and the corresponding functions like `np.multiply`.
+
+This NEPs proposes to refactor the behaviour around two guiding principles:
```
NEPs -> NEP I think

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+2. NumPy scalars or 0-D arrays must always lead to the same behaviour as
+   their N-D counterparts.
+
+We propose to removes all value-based logic and add special handling for
```
`removes` -> `remove`

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+We propose to removes all value-based logic and add special handling for
+Python scalars to preserve some of the convenience that it provided.
+
+This changes also apply to ``np.can_cast(100, np.int8)``, however, we expect
```
`This` -> `These` maybe

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+    to some degree, but this is not currently planned.
+
+
+Impact on operators functions involving NumPy arrays or scalars
```
`operators` -> `operator` maybe

seberg replied:
The `and` was missing, hopefully that makes more sense then :).  Duplicate the example with `np.add` also.

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+------------------------------------------------------------------------
+
+This NEP attempts to preserve most of the convenience that the old behaviour
+gave for Python operators, but remove it for
```
incomplete thought at the end of this sentence perhaps?

seberg replied:
Thanks, mixup there... also a bit of a leftover, because in the first draft, I had angled it for using "weak" logic only on Python operators, which is currently the "alternative".

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+gave for Python operators, but remove it for
+
+The current value-based logic has some nice properties when "untyped" Python
+scalars involved::
```
"are" involved (I think)

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+scalars involved::
+
+  np.arange(10, dtype=np.int8) + 1  # returns an int8 array
+  nparray([1., 2.], dtype=np.float32) * 3.5  # returns a float32 array
```
`np.array`

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+But led to complexity when it came to "unrepresentable" values:
+
+  np.arange(10, dtype=np.int8) + 256  # returns int16
+  nparray([1., 2.], dtype=np.float32) * 1e200  # returns float64
```
`np.array`

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+two examples.  For the second one, the results will be the following::
+
+  np.arange(10, dtype=np.int8) + 256  # raises a TypeError
+  nparray([1., 2.], dtype=np.float32) * 1e200  # warning and returns infinity
```
`np.array`

tylerjereddy replied:
Not sure how much value there is in it, but for the second changed case that returns infinity, are there examples from other frameworks/programming languages that do something similar in a matching scenario?

I don't know that we'd necessarily want to mimic what other major languages/projects do, though citing them as examples may help make a stronger case if needed. Perhaps in this case one could just quickly verify in C or C++ or whatever.

seberg replied:
I added an `admonition`, C normally does this, but of course you can set the evironment to raise an error (just like you can in NumPy, although casts currently don't check that normally!).

I decided to promise to fix that it gives a warning right now.  `pytorch` seems to just overflow silently, I guess they probably don't give warnings reliably, because some hardware they care about doesn't really support it well.

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+Most functions, in particular ``ufuncs`` will also use this weakly typed
+logic.
+In some cases, functions will call `np.asarray()` on inputs before any operations
+and thus will s
```
Incomplete thought here for now I think

seberg replied:
Sorry, this whole paragraph should have been deleted for now, because I moved the option of handling ufuncs differently to the "alternatives".

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+
+Where previously the ``value * 100`` would cause an up-cast to int32/int64
+(because value is a scalar).  The new behaviour will preserve the lower
+precision unless explicitly dealt with (just as if ``value`` was an array).
```
I suppose it would be good to establish that for a number of folks this new behavior would match with the "principle of least surprise." I think the old/current behavior could be convenient for correcting accidents, but perhaps a bit surprising in the context of what one would typically expect for an extracted element of a particular type.

seberg replied:
Lets leave this one open.  I like the "principle of least surprise", but I am not sure I want to burden the backcompat section with it, but rather "advertise" it in the motivation section.

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+Similarliy, if the storage array is float32 a calculation may retain the lower
+float32 precision rather than use the default float64.
+
+Further issues can occure.  For example:
```
`occure` -> `occur`

tylerjereddy commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,443 @@
+
+* `JAX promotion`_ also uses the weak-scalar concept.  However, it makes use
+  of it also for most functions.  JAX further stores the "weak-type" information
+  on the array: ``jnp.array(1)`` remains weakly typed.
```
Heh, the JAX promotion docs are pretty detailed. I was thinking we might want to briefly mention if our new/proposed direction was consistent with some other major languages/frameworks, though not sure if that changes much.

seberg replied:
Yeah, JAX uses the weak logic, I think I should be reporting some bugs to them, but I don't remember (I believe if you swapped the inputs on some of their example, things got wrong, and JAX tags the "was scalar" to the 0-D array, so it can be passed around, which I am not sure always turns out right, but is easier for them because their arrays are immutable.

I believe most use "weak" logic (torch definitely does), although all of these skip scalars entirely probably.

jakevdp commented:
Hi all - @shoyer pointed me to this PR. I wanted to mention that I recently published a detailed design doc outlining the decisions behind JAX's type promotion system, which you can read here: https://jax.readthedocs.io/en/latest/design_notes/type_promotion.html

jakevdp commented on doc/neps/nep-0050-scalar-promotion.rst:
````diff
@@ -0,0 +1,443 @@
+  ```
+* Certain operations are expected to start failing:
+  ```python3
+  np.array([1], np.uint8) * 1000
````
Can you clarify, this fails because `1000` is not representable in `uint8`. But am I understanding correctly that `np.array([1], np.uint8) * 100` would silently overflow, because `100` is representable in `uint8`?

My main area of confusion from the set of examples here is which operations will silently overflow and which will error.

seberg replied:
Yes, that is what I have in mind currently.  I am not at all set on it though – it would actually be much simpler to not do that!
Basically, my rough plan would be that the above would fail, but `np.uint8(1000)` would not for now, so that you could still write `np.uint8(-1)` if you must for some reason – because... backward compatibility mostly.

That is painful, so a simpler middle ground could also be that both `np.uint8(1000)` and `np.uint8(1) + 1000` would give the same warning.

In general, NumPy does not give reliable warnings for integer overflows.  We only give warnings for scalar python operations (and even then randomly, because scalar operator calls are poorly organized).  So `np.uint8(10) + 250` will give a warning, but that example would not (it works with an array).

That is one argument for why things could be a smoother transition if we limit to Python operators:  For those (and scalars which are currently the place that upcasts a lot), we could pull off to ensure warnings are given.  (Super painful because it requires cleaning up the scalar mess, but a clear path.)

One thing that I am unsure about what JAX does, is that I don't really think we should tag on the "weak" property to the array as JAX does.  One, because I am worried it may be brittle/surprising to carry it around for long and second because unlike JAX our arrays are mutable, so there is an additional complexity.  (We can still tag on something as an implementation detail.)

jakevdp replied:
Thanks - regarding JAX & the weak type attribute: the reason we used that approach is that in JAX's JIT-compiled code, everything is represented by a JAX array (scalar inputs get automatically promoted). So to maintain the same behavior between JIT and non-JIT code paths, we need a way to mark a JAX array as weakly-typed. Since NumPy doesn't have that requirement, it seems that it would make sense to avoid it.

That said, one benefit of JAX's approach is that calling `jnp.asarray` on a function's inputs does not change the results even if the input is a scalar, because `asarray` maintains the weakness of the array.

On the other hand, allowing zero-dimensional arrays to be weak readily leads to cases where larger arrays can be weak as well (e.g. `jnp.broadcast_to(jnp.array(1.0), (100, 100))` returns a weakly-typed matrix) so weak types can no longer be thought of as strictly synonymous with scalars. That's a bit surprising, and is one reason we're currently actively discussing design considerations around the semantics of weakly-typed values.

seberg replied:
Ah, that is interesting (and I should mention the weak attribute in alternatives, even if I currently would need some convincing to consider it more seriously – and add 1-2 more examples here for clarity of the initial point, of course).

The broadcast example is nicely clear about how things may get confusing when dragging on the flag.  Long ago, I had early plans to support, in theory, even `[1, 2, 3]` to be considered "weak" – e.g. for ufuncs – but I had abandoned them: it seemed just to much and also way too many moving parts at the time.

In the end, it would be nice to have a clear concept of how far the "weakness" reaches.  I could half imagine an `np.asarray(x, allow_weak=True)` for library use, but I somewhat imagine libraries are better of handling it up-front with `np.result_type` (or by not calling `asarray` at all) – possibly `asarray_unless_literal(x)` if need be.

> That's a bit surprising, and is one reason we're currently actively discussing design considerations around the semantics of weakly-typed values.

It will be interesting to hear about that!  NumPy might end up being a bit more limited in options due to backcompat, but it would be nice if we can manage to align some concepts, i.e. at which step "weakness" is forgotten.

seberg committed:
DOC: Fixups and brief "weakly typed arrays" discussion/alternative

Based on Tylers and Jake Vanderplas' comments.

mhvk commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,468 @@
+
+This would be the simplest solution, however, it would lead to many upcasts when
+working with arrays of ``float32`` or ``int16``, etc.  The solution for these cases
+would be to rely on in-place operations.
```
It may be good to write it as an explicit advantage of the proposed scheme that operations with a python scalar give the same result whether or not it is done in-place.

mhvk reviewed (commented):
This is a great start! I very much like removing distinctions between scalars and 0 or any D arrays.

I also agree that there is least surprised by interpreting python values as literals for which, effectively, one does something like `np.array(python_literal)` and, if possible, does `.astype(self.dtype, casting='same_kind')` (i.e., allow cast from int to float and float to complex, but not the other way around).

It may be good to be explicit about what will happen for:
```
np.arange(10, dtype='i2') * 1.
np.arange(10, dtype='f4') * 1j
```
Surely, the answer should be float and complex, but should there be an attempt to stay as close as possible to the dtype? I.e., `float32` and `complex64`? Especially the latter would seem nice (though my use case would be better served by having an `expj` ufunc!).

mhvk commented:
An alternative perhaps worth discussing (in the considered but rejected section) is that for python scalars values do matter, and that they always are interpreted as the smallest dtype that can hold them. I think that can be rejected given the immediate problems with numbers like `0.1` - making that a `float16` would be pretty bad as `np.arange(10.) + 0.1` would become equivalent to
```
np.arange(10.)+np.float16(0.1)
array([0.09997559, 1.09997559, 2.09997559, 3.09997559, 4.09997559,
       5.09997559, 6.09997559, 7.09997559, 8.09997559, 9.09997559])
```
while on the other hand one would not want to just go for `np.float64` unconditionally. So, sticking to the `dtype` of the array would make more sense.

mhvk commented:
p.s. I realize some of my examples are well answered in @jakevdp's write-up (very nice!). I suggest adding something like the lattice there, as it really helps to visualize the promotions.

seberg commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,468 @@
+
+Will both return a float64 rather than float32.  This improves precision but
+slightly changes results and uses double the memory.
+
```
Note to self:  A weird corner case is that `arr + 10**100` is currently "obviously" an `object` array result.  We either need to decide to preserve this or note here that users may be forced to write `np.asarray(10**100)` to trigger the slightly awkward "use object if necessary" logic – or provide `dtype=object`.

jni replied:
I think automatic promotion to dtype object is always a bad idea. Why would this not just overflow?

seberg replied:
Added a subsection to the backwards compatibility part (below), so "resolving" for now

seberg commented:
@rgommers I did not dig deep yet, but I think the vast majority of failures fall into three categories:
* The tests simply need adjusting: they are thorough and check for result types, but result types change.  I think (I am not quite sure!) that the sparse tests fall into this category (at least mainly).
* A few floating point equality (or approximate equality) checks fail because `np.float32(...) == 123.32123` casts the Python float to float32 (it is considered "weak") rather than vice-versa, and that requires adaptation of test sensitivity.
* E.g. integrators sometimes pass on scalars, if the input is float32 the inner function would have ended up using float64 anyway (due to the upcast of scalars).  If that changes, that means some results have reduced precision and tests fail.

Right now I expect most of these are tedious, but not particularly problematic to resolve for SciPy (and many other libraries).  The impact seems much more on the end-user side rather than for libraries.

However, I am not sure if SciPy or other libraries will want to support Python scalars as arguments (without calling `np.asarray(argument)` indiscriminately).  If libraries want that, this may be a very important user-story to figure out and propose a solution for in the NEP.  (I don't think it is hard to find a solution, but I am not sure the exact direction to go right now.)

seberg committed:
NEP: Revise and extend abstract+motivation and add a promotion schema

seberg committed:
NEP: Insert Backward compatibility subsection on "integer ladder" with object

This may be a bit "too much" information, but on the other hand it
is a backcompat break

seberg committed:
Add sentence noting "principle of least surprise" as suggested by Tyler

seberg committed:
DOC: Add additional examples early in the NEP and fix the "category" promotion

Even if I prefer not to mention the word "category" too much, integral vs. inexact
is the "magic" boundary when it comes to Python promoting an numpy
scalar to the higher category.

seberg committed:
DOC: Add rejected "alternative" of using value inspection on Python scalars

rgommers commented:
> Well, I am thinking of "ufunc-like" functions mainly, I guess.

Ah okay, then I can't think of any right now - such functions accept scalars only as array-like and call `asarray` on them I expect.

seberg committed:
DOC: Fix schema, accidentally safed it under the wrong name

stefanv committed:
Styling edits

rossbar committed:
minor grammatical/rst formatting edits.

rossbar reviewed (commented):
Thanks for putting this together @seberg . I left some inline comments/suggestions from a first readthrough. I'm going to let it marinate then reread it again!

seberg committed:
NEP: Add comparison table and "old behaviour" explanation.

But I need to link to the old behaviour explenatation early on
somewhere...
And possibly it is now too much.  Plus, I am not sure the table
is actually all *that* useful

seberg commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,601 @@
+
+After the change, the promotions in NumPy will follow the schema below.
+Promotion always occurs along the green lines:
+from left to right within their kind and to a higher kind only when
```
You are right, will move the bubble.  It is super annoying, because technically, the bubble should be to the left of `uint64/int64`, but for practical reasons, numpy chooses to not put it there.
But... moving the bubble is right in practice for that reason.

seberg commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,601 @@
+   :ref:`NEP 42 <NEP42>`.
+   Currently, this leads to a dual implementation of a new and an old (value
+   sensitive) system.  Fixing this will greatly simplify the internal logic
+   and make results more consistent.
```
Well, technically the current system is consistent between arrays and scalars.  It is inconsistent between 0-D and not 0-D, so not sure...

seberg committed:
NEP: Update NEP 50 promotion schema (and other small fixes)

As per comment by Ross.  float64 can be put at the same precision
as "uint64".  This is incorrect, but it is the practical choice
that NumPy uses.

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,601 @@
+
+    np.array([100], dtype=np.uint8) + 200 == np.array([44], dtype=np.uint8)
+
+Considering this, we believe that the proposal follows the "principle of least
```
I don't understand this, as the table above shows that this behaviour is maintained?

Silent integer overflow is a viciously surreptitious source of bugs, and if I read the NEP correctly, we're increasing the surface for that by defaulting to lower integer types - e.g. `uint8(1) + 2`: `int64(3)` (old) -> `uint8(3)` (new)

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,729 @@
+minimum/default precision: that is ``float64``, ``complex128`` or ``int64``
+(``int32`` is used on some systems, e.g. windows).
+
+.. figure:: _static/nep-0050-promotion-no-fonts.svg
+    :figclass: align-center
```
The horizontal lines in this graphic should have the same thickness and colour as the diagonal ones. At first glance it looks like `uint8` can only promote to `float16`, not `uint16`.

![grafik](https://user-images.githubusercontent.com/33685575/163495919-987c3816-0ce0-496e-abaa-e9c9263ea13f.png)

seberg replied:
Ah, got some alpha wrong or so, will fix, thanks.

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,729 @@
+  # and:
+  np.add(np.arange(10, dtype=np.uint8), np.int64(1))
+
+Will return an int64 array because the type of ``np.int64(1)`` is strictly
+honoured.
```
It's not clear from context ("currently surprising cases") if this "will" is describing current or future behaviour.

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,601 @@
+more strict:  This means that the results will be unchanged in the first
+two examples.  For the second one, the results will be the following::
+
+  np.arange(10, dtype=np.int8) + 256  # raises a TypeError
```
Agreed with @rossbar. This is likely to break a ton of essentially harmless code. I think it would be great to raise on integer overflow (consistently, not just with python scalars), but I think it would be vastly preferable to only do that when we actually fall out of (u)int64.

It's laudable to be more precise on this (i.e. if someone _really_ wants to make sure their `int8` stays an `int8`), but I think it's going to be a very sharp edge for the large majority of users who do not care about the underlying representation of their ints, but will be hit with errors much more often.

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,729 @@
+
+.. admonition:: Behaviour in other libraries
+
+   Overflowing in the conversion rather than raising an error is a choice;
+   it is one that is the default in most C setups (similar to NumPy C can be
+   set up to raise an error due to the overflow, however).
```
I strongly [believe](https://github.com/numpy/numpy/issues/21219#issuecomment-1073184844) that we should never silently overflow, but promote to the end of the integer hierarchy before doing that (so if there's ever a portable and universally available `int128`, we should keep going IMO).

I realise that this is incompatible with the proposed move away from value-based promotions, but to me that seems a step in the wrong direction (not talking about the internals and their complexity, but from a UX POV).

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,729 @@
+Where previously the ``value * 100`` would cause an up-cast to int32/int64
+(because value is a scalar).  The new behaviour will preserve the lower
+precision unless explicitly dealt with (just as if ``value`` was an array).
+This can lead to integer overflows and thus incorrect results beyond precision.
+In many cases this may be silent, although NumPy usually gives warnings for the
+scalar operators.
```
This sound like a really bad idea to me. I'm sorry to be so negative on this, but this could well lead to silent and catastrophic failures & data corruption, and IMO poses a substantial risk of blemishing numpy's reputation if not handled with extreme care.

seberg replied:
Note that I am planning to add at least an *optional* warning tthat will notify of places where reduced precision occurs (I am not sure it would be 100%, but it should be good enough in practice).

stefanv replied:
Why will these warnings be optional?  This is the case I am by far the most worried about.

seberg replied:
Note that there are two warnings involved here:
* Overflow warnings are normally given for NumPy scalar operations (which are the main affected ones here).  (They are not given for array operations.)
* Warnings about the actual change, which I was thinking should probably be opt-in.

For the second, I was aiming for opt-in currently, because I think they will be very noisy and cumbersome to avoid.  E.g. for `float32 >= 2.1` where the behaviour might change, but practically never does.  Should we do the operation twice to check? Give a warning when it probably doesn't matter?

Maybe we could check whether always warnings should happen for more particularly surprising changes, right now I hoped the scalar integer overflow warnings cover the worst of it though.

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,729 @@
+* Certain operations are expected to start failing::
+
+     np.array([1], np.uint8) * 1000
+     np.array([1], np.uint8) == 1000  # possibly also
```
Why "possibly"? What does it depend on?

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,729 @@
+
+In other cases, increased precision may occur.  For example::
+
+  np.multiple(float32_arr, 2.)
+  float32_arr * np.float64(2.)
+
+Will both return a float64 rather than float32.  This improves precision but
```
I thought python scalars would be treated as "weak"? Why does the first one upcast to float64?

seberg replied:
I think I use "Python scalars" for `int`, `float`, `complex`.  So the ones builtin with Python, not the *NumPy* scalars.  But maybe I can try to avoid that term to avoid the tricky distinction.

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,729 @@
+but it would cement a distinction between arrays and scalars (NumPy arrays
+are "stronger" than Python scalars, but NumPy scalars are not).
+
+Such a distinction is very much possible, however, at this time NumPy will
+often (and silently) convert 0-D arrays to scalars.
```
To me, it's more important that 0-D arrays behave consistently to their N-D counterparts than that they behave consistently with scalars. Some libraries such as CVXPY have tightened this distinction in recent times, and I think this is a healthy choice.

A numpy 0-D array has a completely different API than a python scalar, and liberally casting one to the other is not beneficial in my opinion. I think they should be _compatible_, but that types should be maintained, e.g. I expect a function operating on scalar values to maintain the input type (0-D array -> 0-D array; python scalar -> python scalar; without silently changing between them).

seberg replied:
Yes, but NumPy does not currently have that choice, because we create scalars from 0-D arrays silently all the time.  If scalars were distinct objects, you only got with a special operation.  I might agree.

This NEP is only about the *dtype* of the array or scalar, not about this behaviour.  I agree that it should be cleaned up, but it is another set of worms.  But I also think that NumPy scalars have a `dtype` and because of that should behave the same as NumPy arrays for promotion.

h-vetinari commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,729 @@
+to NumPy scalars and 0-D arrays, rather than the application to Python scalars.
+We could thus consider to keep inspecting the value for Python scalars.
+
+We reject this idea on the grounds that it will not remove the surprises
+given earlier::
+
+    np.uint8(100) + 1000 == np.uint16(1100)
+    np.uint8(100) + 200 == np.uint8(44)
```
I'd argue that the first one is fine and the second should be `np.uint16(300)` (more details in the comments above).

h-vetinari reviewed (commented):
I have a queasy feeling about this. I cannot speak about the benefits to improvements of implementation complexity of this proposal (though I don't doubt they are substantial), but from a user experience point it seems to be going into the wrong direction IMO.

If anything, I'd [argue](https://github.com/numpy/numpy/issues/21219#issuecomment-1073184844) we should be fighting silent integer overflow tooth and nail, not introducing more surface where it might appear (e.g. by tending to default more to smaller integer types). I consider the breaking changes around this to be a non-negligible risk to numpy's reputation.

Moving away from the (value-based) promotion through the integer hierarchy is most likely similarly disadvantageous for the large majority of users who do not care about the machine representation of their integers.

I also tend to think that the distinction between 0-D arrays and scalars should be sharpened rather than weakened. IMO arrays should behave consistently (regardless of dimensionality), but the APIs between numpy arrays and python scalars are so different that silently ending up with the other one also poses several sharp edges.

leofang commented:
cc: @kmaehashi @toslunar

seberg commented:
@h-vetinari I have to catch up with this, I do not have a lot of time right now.

You seem to be mainly worried about integer overflows and the behaviour of that.  The reality of it is that for *array* operations, we never check values.  Most importantly we do *not* check the values of results.  I do have an example somewhere that:
```
np.array([100], dtype=np.uint8) + 300  ->  uint16(400)
np.array([100], dtype=np.uint8) + 250  ->  uint8(94)
```
So we are *increasing* the amount of places where this happens, that is true.  But the situation is much less useful currently than it may appear.
NumPy should not (and technically canont well) look at all values of two arrays.  We also cannot really look at what the *result* would be to decide the output `dtype`.  Our `uint8` and `int8` are **not** "integers" that are stored as `int8`, they are actual `int8` in the C sense (adding two overflows in some way).  They cannot be, since that would require to either just make the result `int64` always (that does work but is not the reality [1]) or dynamically resize the dtype?

In practice, I do not think going the other way is a technically possible alternative at all.  IMO NumPy has.

Should I add this as an additional sub-section to the (rejected) alternatives?

---

[1] In theory, we could have an `int64[int8]` on day.  But the logic for it would be that it always returns an `int64` on value-changing operations.

h-vetinari commented:
> Most importantly we do _not_ check the values of results. I do have an example somewhere that:
> 
> ```
> np.array([100], dtype=np.uint8) + 300  ->  uint16(400)
> np.array([100], dtype=np.uint8) + 250  ->  uint8(94)
> ```

This to me is one of the most surprising and dangerous parts, and IMO an aspect of the design that needs to be fixed. Imagine that your code gets exercised during testing only with values falling into the first line, and once you're in production, you suddenly (and silently) hit the second one.

I think the idea to treat python scalars to only influence the type "weakly" is sound, but then it needs to take into account the value it is being added to to avoid silently causing overflow by choosing a too-small type for said python scalar in a _really_ innocuous piece of code.

> Our `uint8` and `int8` are **not** "integers" that are stored as `int8`, they are actual `int8` in the C sense (adding two overflows in some way).

I understand that. My point is we should strive to protect users from the problems of C integers, rather than expose them even more.

> We also cannot really look at what the _result_ would be to decide the output `dtype`.

I think it's come up in various different shapes already, but would it make sense to expose different modes/errstates/etc. that allow users to explicitly opt into performance (no overflow check), while the default could be that we _do_ take the performance hit to introspect enough to warn/error on overflow.

shoyer commented:
> > We also cannot really look at what the _result_ would be to decide the output `dtype`.
> 
> I think it's come up in various different shapes already, but would it make sense to expose different modes/errstates/etc. that allow users to explicitly opt into performance (no overflow check), while the default could be that we _do_ take the performance hit to introspect enough to warn/error on overflow.

I agree that support for explicitly checking for numerical overflow would be a great option to have, similar to how we handle dividing by zero (error, warn or ignore).

I'm not sure which default behavior I would prefer (maybe a warning would be a good compromise?), but any of these would be better than the current value dependent dtype promotion behavior.

seberg commented:
> This to me is one of the most surprising and dangerous parts, and IMO an aspect of the design that needs to be fixed. Imagine that your code gets exercised during testing only with values falling into the first line, and once you're in production, you suddenly (and silently) hit the second one.

And to what?  It is difficult and currently impossible to take into account *all* values (so the values of the non-scalar).  It is also impossible from a typing perspective if you want to type dtypes/scalar types.

My plan is to raise an error for integer overflows here where we currently upcast.  I think this is the better behaviour, you get a loud failure when "weird things" may be happening.  The case where upcasting does *not* happen but would be necessary to avoid overflow is *unchanged* by this proposal.
Obviously, it would be nice to have optional warnings for all integer overflows, but I don't really see that it matters for this NEP: The status-quo is unchanged or improved when it comes to warning/errors for these cases.

seberg committed:
DOC: Fix line width in svg

seberg committed:
DOC: Small NEP updates based on comments

seberg committed:
DOC: Small type-edit style changes and rewordings

stefanv commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,743 @@
+Usage and Impact
+================
+
+This NEP is expected to be implemented with **no** transition period that warns
```
Yikes; this has the potential to break some codes out there, without any obvious way to fix them.  We should probably, at least, make a major version release.

stefanv commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,743 @@
+
+The main impact on operations not involving Python scalars (``float``, ``int``,
+``complex``) will be that 0-D arrays and NumPy scalars will never behave
+value-sensitive.
```
```suggestion
value-sensitively.
```

Although, this phrasing is slightly odd; can reword perhaps to something like "that operations on 0-D arrays and NumPy scalars will never depend on their values"

stefanv commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,743 @@
+Impact on operators involving Python ``int``, ``float``, and ``complex``
+------------------------------------------------------------------------
+
+This NEP attempts to preserve the convenience that the old behaviour
+gave when working with literal values.
```
```suggestion
This NEP attempts to preserve the convenience of the old behaviour
when working with literal values.
```

stefanv commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,743 @@
+The proposal is to preserve this behaviour for the most part.  This is achieved
+by considering Python ``int``, ``float``, and ``complex`` to be "weakly" typed
+in operations.
+However, to avoid user surprises, we plan to make conversion to the new type
```
```suggestion
However, to avoid surprises, we plan to make conversion to the new type
```

stefanv commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,743 @@
+  np.array([1., 2.], dtype=np.float32) * 1e200  # warning and returns infinity
+
+The second one warns because ``np.float32(1e200)`` overflows to infinity.
+It will then do the calculation with ``inf`` as normally.
```
```suggestion
It will then continue to do the calculation with ``inf`` as usual.
```

stefanv commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,743 @@
+Motivation and Scope
+====================
+
+The motivation for changing the behaviour with respect to inspecting the value
```
I see this is mentioned below; perhaps a forward reference to that section.

seberg replied:
Adding a third item in the "motivation", that this will align us with other projects (and then link to there)

seberg committed:
Apply suggestions from code review

Co-authored-by: Stefan van der Walt <sjvdwalt@gmail.com>

seberg commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,743 @@
+from left to right within their kind and to a higher kind only when
+necessary.
+The result kind is always the largest kind of the inputs.
+Note that NumPy considers ``float32`` to have a lower precision than ``int32``
```
The reason is that float32 cannot exactly represent all int32 values, so it makes sense.  The implication in the figure is that I moved float32 slightly to the left of int32.

It is not really safe to cast int32 to float32 or int64 to float64 and because of that promotion avoids int32 to float32 (it goes to float64).  But we make the exception that we consider int64 to float64 a safe cast and a valid promotion (even though it arguably loses precision).

seberg replied:
The implication on the figure is that the float32 bubble is to the left of the in32 to bubble

seberg committed:
NEP: update wording to clarify that 255 requires an int16 not int8

seberg commented on doc/neps/nep-0050-scalar-promotion.rst:
```diff
@@ -0,0 +1,743 @@
+Note that the schema omits, ``bool``.  It is set below "integral", so that the
+following hold::
+
+    np.bool_(True) + 1 == np.int64(2)
```
Hmmm, certainly also plausible.  My gut feeling is that this is the right choice, because:
```
np.int8(3) + 1.
```
should be float64 I think, since there is no clear logic to prefer float32 (or float16?).

seberg committed:
NEP: Address more of Stéfan's review comments

seberg commented:
The next larger iteration will be to create a branch (or PR) that doesn't just show the new behaviour, but also implements all the warnings I would like to have.
That may be necessary to really discuss how bad the break is for downstream.

stefanv reviewed (approved):
Agreed that we should merge the draft NEP, and then start the approval process.

seberg committed:
NEP: Add link to NEP50 (useful for intersphinx even if unused)

seberg committed:
NEP: Update "created" date to date draft goes in (or closer)

mattip commented:
Putting this in not to end discussion but to provide a basis for further discussion

seberg commented:
Indeed, it is just a basis for further discussion!  @jakevdp, I am very curious whether there has been any further development/thoughts along this in the JAX team?  Since I think you may have some important insights for NumPy as well.
